File: src/test/java/uk/co/agware/carpet/TestScriptTask.java
package uk.co.agware.carpet;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import uk.co.agware.carpet.change.tasks.ScriptTask;
import uk.co.agware.carpet.database.DatabaseConnector;

/**
 * Created by Philip Ward <Philip.Ward@agware.com> on 14/06/2016.
 */
public class TestScriptTask {

    private DatabaseConnector databaseConnector;

    @Before
    public void buildMocks(){
        databaseConnector = Mockito.mock(DatabaseConnector.class);
        Mockito.when(databaseConnector.executeStatement(Mockito.anyString())).thenReturn(true);
    }

    @Test
    public void testSimpleConstructor(){
        ScriptTask task = new ScriptTask("Task 1", 1, "SELECT * FROM Table", ";");
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        Mockito.verify(databaseConnector, Mockito.times(1)).executeStatement("SELECT * FROM Table");
    }

    @Test
    public void testSimpleConstructorNoDelimiter(){
        ScriptTask task = new ScriptTask("Task 1", 1, "SELECT * FROM Table", "");
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        Mockito.verify(databaseConnector, Mockito.times(1)).executeStatement("SELECT * FROM Table");
    }

    @Test
    public void testSimpleConstructorNullDelimiter(){
        ScriptTask task = new ScriptTask("Task 1", 1, "SELECT * FROM Table", null);
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        Mockito.verify(databaseConnector, Mockito.times(1)).executeStatement("SELECT * FROM Table");
    }

    @Test
    public void testMultipleStatements(){
        StringBuilder script = new StringBuilder("SELECT * FROM Table;");
        script.append("SELECT * FROM another_table;");
        script.append("DELETE FROM TABLE");
        ScriptTask task = new ScriptTask("Task 1", 1, script.toString(), null);
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        ArgumentCaptor<String> statements = ArgumentCaptor.forClass(String.class);
        Mockito.verify(databaseConnector, Mockito.times(3)).executeStatement(statements.capture());
        Assert.assertEquals(3, statements.getAllValues().size());
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Table"));
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM another_table"));
        Assert.assertTrue(statements.getAllValues().contains("DELETE FROM TABLE"));
    }
}

File: src/test/kotlin/uk/co/agware/carpet/TestScriptTask.kt
package uk.co.agware.carpet

import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.*
import org.junit.platform.runner.JUnitPlatform
import org.junit.runner.RunWith
import uk.co.agware.carpet.change.tasks.ScriptTask
import uk.co.agware.carpet.database.DatabaseConnector
import uk.co.agware.carpet.exception.MagicCarpetParseException
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

@RunWith(JUnitPlatform::class)
class TestScriptTask: Spek(spek)
private val spek: Dsl.() -> Unit = {

    describe("A ScriptTask object") {

        var connection = mock<DatabaseConnector>()

        beforeEachTest {
            connection = mock<DatabaseConnector>()
        }

        given("A Script Task with a default delimiter") {
            val subject = ScriptTask("Test Task", 1, "SELECT * FROM Table; SELECT * FROM Other_Table")

            on("Performing the task") {

                subject.performTask(connection)

                it("should execute each statement") {
                    val statement = argumentCaptor<String>()
                    verify(connection,times(2)).executeStatement(statement.capture())
                    assertEquals(2, statement.allValues.size)
                    assertTrue(statement.allValues.contains("SELECT * FROM Table"))
                    assertTrue(statement.allValues.contains("SELECT * FROM Other_Table"))
                }
            }
        }

        given("a script task with a custom delimiter") {
            val subject = ScriptTask("Test Task", 1, "SELECT * FROM Table, SELECT * FROM Other_Table", ",")

            on("Performing the task") {

                subject.performTask(connection)

                it("should execute each statement") {
                    val statement = argumentCaptor<String>()
                    verify(connection,times(2)).executeStatement(statement.capture())
                    assertEquals(2, statement.allValues.size)
                    assertTrue(statement.allValues.contains("SELECT * FROM Table"))
                    assertTrue(statement.allValues.contains("SELECT * FROM Other_Table"))
                }
            }
        }

        it("should fail with a MagicCarpetParseException") {
            assertFailsWith<MagicCarpetParseException> {
                ScriptTask("Test Task", 1, "")
            }
        }
    }

}
